qiskit_experiments/analysis/curve_fitting.py

```diff
@@ -0,0 +1,304 @@
+
+    for multiple series of :math:`x_k, y_k, \sigma_k` data evaluated using
+    a list of objective functions :math:`[f_k]`
+    using ``scipy.optimize.curve_fit``.
+
+    Args:
+        funcs: a list of objective functions :math:`[f_0, f_1, ...]` where
+               each function has signature :math`f_k`(x, *params)`.
+        series: a 1D int array that specifies the component objective
+                function :math:`f_k` to evaluate corresponding x and y
+                data with.
+        xdata: a 1D float array of xdata.
+        ydata: a 1D float array of ydata.
+        p0: initial guess for optimization parameters.
+        sigma: Optional, a 1D array of standard deviations in ydata
+               in absolute units.
+        weights: Optional, a 1D float list of weights :math:`w_k` for each
+                 component function :math:`f_k`.
+        bounds: Optional, lower and upper bounds for optimization
+                parameters.
+        kwargs: additional kwargs for scipy.optimize.curve_fit.
+
+    Returns:
+        result containing ``popt`` the optimal fit parameters,
+        ``popt_err`` the standard error estimates popt,
+        ``pcov`` the covariance matrix for the fit,
+        ``reduced_chisq`` the reduced chi-squared parameter of fit,
+        ``dof`` the degrees of freedom of the fit,
+        ``xrange`` the range of xdata values used for fit.
+
+    Raises:
+        QiskitError: if the number of degrees of freedom of the fit is
+                     less than 1.
+
+    .. note::
+        ``sigma`` is assumed to be specified in the same units as ``ydata``
+        (absolute units). If sigma is instead specified in relative units
+        the `absolute_sigma=False` kwarg of scipy curve_fit must be used.
+        This affects the returned covariance ``pcov`` and error ``popt_err``
+        parameters via ``pcov(absolute_sigma=False) = pcov * reduced_chisq``
+        ``popt_err(absolute_sigma=False) = popt_err * sqrt(reduced_chisq)``.
+    """
+    num_funcs = len(funcs)
+
+    # Get positions for indexes data sets
+    series = np.asarray(series, dtype=int)
+    idxs = [series == i for i in range(num_funcs)]
+
+    # Combine weights and sigma for transformation
+    if weights is None:
+        wsigma = sigma
+    else:
+        wsigma = np.zeros(ydata.size)
+        if sigma is None:
+            for i in range(num_funcs):
+                wsigma[idxs[i]] = 1 / np.sqrt(weights[i])
+        else:
+            for i in range(num_funcs):
+                wsigma[idxs[i]] = sigma[idxs[i]] / np.sqrt(weights[i])
+
+    # Define multi-objective function
+    def f(x, *params):
+        y = np.zeros(x.size)
+        for i in range(num_funcs):
+            xi = x[idxs[i]]
+            yi = funcs[i](xi, *params)
+            y[idxs[i]] = yi
+        return y
+
+    # Run linearized curve_fit
+    analysis_result = curve_fit(f, xdata, ydata, p0, sigma=wsigma, bounds=bounds, **kwargs)
+
```

Comment:
The code of this function is difficult to follow in the lack of internal documentation. Also, in the doc string, the meaning of the `series` argument is not clear. The text at the beginning of the doc string implies that `xdata`, `ydata`, and `sigma` are expected to be 2D arrays, but then it turns out that they are 1D arrays, and the connection is not clear.